plugins/producer_plugin/include/eosio/producer_plugin/block_timing_util.hpp

```diff
@@ -1,13 +1,40 @@
 #pragma once
 #include <eosio/chain/block_timestamp.hpp>
 #include <eosio/chain/config.hpp>
+#include <eosio/chain/producer_schedule.hpp>
 
 namespace eosio {
 
 enum class pending_block_mode { producing, speculating };
 
 namespace block_timing_util {
 
+   // Store watermarks
+   class producer_watermarks {
+   public:
+      void consider_new_watermark(chain::account_name producer, uint32_t block_num, chain::block_timestamp_type timestamp) {
+         auto itr = _producer_watermarks.find(producer);
+         if (itr != _producer_watermarks.end()) {
+            itr->second.first  = std::max(itr->second.first, block_num);
+            itr->second.second = std::max(itr->second.second, timestamp);
+         } else {
+            _producer_watermarks.emplace(producer, std::make_pair(block_num, timestamp));
+         }
+      }
+
+      using producer_watermark = std::pair<uint32_t, chain::block_timestamp_type>;
+      std::optional<producer_watermark> get_watermark(chain::account_name producer) const {
+         auto itr = _producer_watermarks.find(producer);
+
+         if (itr == _producer_watermarks.end())
+            return {};
+
+         return itr->second;
+      }
+   private:
+      std::map<chain::account_name, producer_watermark> _producer_watermarks;
+   };
+
    // Calculate when a producer can start producing a given block represented by its block_time
    //
    // In the past, a producer would always start a block `config::block_interval_us` ahead of its block time. However,
```

Comment:
Hum, I don't quite understand this comment. Suppose we have 12 repetitions. given block_interval=500 ms and cpu effort=400 ms, assuming the our round start at time point 0; in the past, the block start time points would be at time point : `-500, 0, ... , 5000, 5500`, so the last block would start as expected in the 12th slot.

With the new approach, the block time points would become `-500, -100, 300, 700, 1100 ..., 3900, 4300`. Does this mean we start our last block at 4.3s and that we process no more transactions for 1.7s until the next producer schedule starts?

Reply:
Yes. See https://github.com/AntelopeIO/leap/pull/867#issuecomment-1481596497

Reply:
Sounds odd that we guarantee having 1.1s gaps where no transactions are included in blocks and no blocks produced.

Reply:
Oh, never mind, I forgot this was only for non-producer nodes.

Reply:
It is for production nodes.

Reply:
So isn't it odd that we guarantee having 1.1s gaps where no transactions are included in blocks and no blocks produced?